exercise/exercise_models.py

```diff
@@ -275,6 +275,32 @@ def get_templates(self):
         entries = pick_localized(self.templates, get_language())
         return [(url,url.split('/')[-1]) for url in entries.split()]
 
+    def intersect_form_spec_keys(self, another_lobject, include_static_fields=False):
+        """Return the intersection of form field keys from this learning object and another.
+        In other words, keys that exist in both forms. The form keys are read
+        from the form_spec list of the exercise_info JSON field.
+        """
+        if not (self.exercise_info and another_lobject.exercise_info \
+                and 'form_spec' in self.exercise_info \
+                and 'form_spec' in another_lobject.exercise_info):
+            return set()
+
+        def make_key_type_pairs(form_spec_list):
+            pairs = set()
+            for item in form_spec_list:
+                key = item.get('key')
+                typ = item.get('type')
+                if not include_static_fields and typ == 'static':
+                    continue
+                if key and typ: # avoid empty or missing values
+                    pairs.add((key, typ))
+            return pairs
+
+        keys1 = make_key_type_pairs(self.exercise_info['form_spec'])
+        keys2 = make_key_type_pairs(another_lobject.exercise_info['form_spec'])
```

Comment:
Notice that `form_spec` might be missing, in which case, I would presume an empty dictionary, In addition, `exercise_info`, might not be a dictionary (it's what ever json data, e.g. list or just a string).

Maybe `spec = self.exercise_info.get('form_spec', {}) if isinstance(self.exercise_info, dict) else {}`

Reply:
This is mostly irrelevant, relevant commit: https://github.com/Aalto-LeTech/a-plus/pull/393/files#r313386334